Support window.onload

kt3k opened the pull request:
This PR tries to solve #2386. I added `triggerLoadEvent` function for triggering `window.onload` function (I put it under denoMain namespace to avoid leaking to global namespace). `run_script` function (in //cli/mian.rs) calls it after the `main_module` executed.

---
closes #2386

bartlomieju commented on tests/034_onload.ts:
```diff
@@ -0,0 +1,4 @@
+window.onload = function(e): void {
+  console.log(`got ${e.type} event`);
+};
+console.log("main");
```
Might be good idea to add 1-2 imports in this file, that do `console.log`  to ensure that `onload` is actually called after imports are resolved. WDYT?

kt3k replied:
Agree. I added an import statement.

kevinkassimo commented on cli/main.rs:
```diff
@@ -314,6 +314,7 @@ fn run_script(flags: DenoFlags, argv: Vec<String>) {
     worker
       .execute_mod_async(&main_module, false)
       .and_then(move |()| {
+        js_check(worker.execute("denoMain.triggerLoadEvent()"));
```
I feel we might want to give it a more general name (I am thinking that we might add more lifecycle hooks at some point)

kt3k replied:
Changed to `denoMain.lifecycles.onLoad` for now. Maybe we can add more lifecycles in this namespace. I'm open to any further suggestion!

kevinkassimo commented on js/main.ts:
```diff
@@ -51,3 +52,5 @@ export default function denoMain(name?: string): void {
   }
 }
+
+denoMain.triggerLoadEvent = triggerLoadEvent;
```
I think we need to at some point collect all private calls (e.g. `window.workerMain`, `window.xevalMain`) and this to a centralized location to avoid confusing users with other public ones. Maybe putting into `window._$$deno` so that they look like `window._$$deno.triggerLoadEvent`? (debating about the naming)

kt3k replied:
I was a little uncomfortable about using denoMain as a namespace. I agree that we should have something like what you suggest.

ry commented:
@kt3k Nice!

In the browser you can also do something like `window.addEventListener("load", x)`. I think that would be nice to support here.

kt3k commented:
@ry OK. I'll try that. Probably we can reuse the EventTarget implementation.

ry commented on js/window.ts:
```diff
@@ -8,3 +7,23 @@ export const window = (0, eval)("this");
+    onload(e);
+  }
+});
+*/
```
(remove before landing)

ry commented on js/event_target.ts:
```diff
@@ -525,31 +550,11 @@ export class EventTarget implements domTypes.EventTarget {
-  enumerable: true,
-  writable: true
-});
 Reflect.defineProperty(EventTarget.prototype, "addEventListener", {
```
I'm having trouble understanding all the changes in event_target.ts - what is happening here?

kt3k replied:
I think true public interfaces of EventTarget are only 3 methods (addEventListener, removeEventListener, and dispatchEvent), and the other methods and properties (host, mode, nodeType, etc) are implementation details for this specific implementation. Here we want to make `window` an EventTarget, but if we naively put all those non-public methods and properties to window, they all become global variables. To avoid that, I refactored EventTarget to use symbol-keyed properties (like this.host -> this[eventTargetHost]) and locally scoped helper methods (`eventTargetHelpers`). I think these tricks prevent leaking things to the global namespace.

ry commented on tests/034_onload/main.ts:
```diff
@@ -1,7 +1,14 @@
+  "load",
+  (e: Event): void => {
+    console.log(`got ${e.type} event in event handler (main)`);
+  }
```
:D

kt3k commented:
I changed the several things since the initial version ( diff: https://github.com/denoland/deno/pull/2643/commits/26d6a8907ef022e6810ca65b05b546a930ca49ef )

- Made `window` an EventTarget
- Removed `//js/lifecycle.ts` and `onLoad` function, instead registered the event listener to window which invokes `window.onload`
- Dispatches `load` event after main_module loaded
- Refactored EventTarget to prevent leaking its properties as global variable when making `window` an EventTarget.

ry commented:
@kt3k Looks great! 
Would you mind adding an example to the website/manual.md showing the usage of window.onload ?
